airflow/jobs/scheduler_job_runner.py

```diff
@@ -1838,12 +1890,61 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
     def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[TaskInstance]]:
         """Organize TIs into lists per their respective executor."""
         _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
-        executor: str | None
         for ti in tis:
-            if ti.executor:
-                executor = str(ti.executor)
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                _executor_to_tis[executor_obj].append(ti)
             else:
-                executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            if TYPE_CHECKING:
+                # All executors should have a name if they are initted from the executor_loader. But we need
+                # to check for None to make mypy happy.
+                assert executor.name
+            executor_to_slots_available[executor.name] = executor.slots_available
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                if TYPE_CHECKING:
+                    # All executors should have a name if they are initted from the executor_loader. But we
+                    # need to check for None to make mypy happy.
+                    assert executor_obj.name
```

Comment:
i think these type checking blocks can be removed if we change the typing on `name` attr to be `ExecutorName` instead of `ExecutorName | None`.  i tried to PR into your branch but it got auto closed https://github.com/aws-mwaa/upstream-to-airflow/pull/7

Reply:
Good morning! Sorry, we don't accept any form of PR to our fork. It's strictly for development and upstreaming.

I think I'd like to push back on this one, if you don't mind. Depending on how you're using executors or loading them that field can be None and the typing should reflect that for developers to be aware of it.

Reply:
don't have a strong opinion about it